python/singa/autograd.py

```diff
@@ -1,5 +1,447 @@
+        else:
+            pad = padding
+
+        if dilation != 1 or groups != 1:
+            raise ValueError('Not implemented yet')
+
+        self.PyLayer = layer.Conv2D(inner_params['name'], nb_kernels=out_channels, kernel=kernel_size, stride=stride,
+                                    border_mode=inner_params['border_mode'],
+                 cudnn_prefer=inner_params['cudnn_prefer'], workspace_byte_limit=inner_params['workspace_byte_limit'],
+                 data_format=inner_params['data_format'], use_bias=bias, W_specs=self.W_specs, b_specs=self.b_specs,
+                 pad=pad, input_sample_shape=inner_params['input_sample_shape'])
+
+    def get_params(self):
+        assert self.has_setup, \
+            'Must call setup() before get_params()'
+        params = self.PyLayer.layer.param_values()
+        return params
+
+    def __call__(self, x, flag=True):
+        assert type(flag) is bool, 'flag can only be bool.'
+        if flag:
+            self.flag = model_pb2.kTrain
+        else:
+            self.flag = model_pb2.kEval
+
+        if not self.PyLayer.has_setup:
+            self.PyLayer.setup(x.shape[1:])
+
+        param_data = self.PyLayer.layer.param_values()
+
+        if not hasattr(self, 'w'):
+            self.w = Tensor(device=param_data[0].device, data=param_data[0], requires_grad=True, stores_grad=True)
+            std = math.sqrt(2.0/(self.in_channels*self.kernel_size[0]*self.kernel_size[1]+self.out_channels))
+            self.w.gaussian(0.0, std)
+
+        xs = [x, self.w]
+
+        if len(param_data) == 2:
+            if not hasattr(self, 'b'):
+                self.b = Tensor(device=param_data[1].device, data=param_data[1], requires_grad=True, stores_grad=True)
+                self.b.set_value(0.0)
+
+            xs.append(self.b)
+
+        xs = tuple(xs)
+        return self._do_forward(*xs)[0]
+
+    def forward(self, *xs):
+        return self.PyLayer.layer.Forward(self.flag, xs[0])
+
+    def backward(self, dy):
+        ret = self.PyLayer.layer.Backward(self.flag, dy)
+        return (ret[0],)+ret[1]
+
+
+class Linear(Operation):
+    def __init__(self, in_features, out_features, bias=True):
+        self.in_features = in_features
+        self.out_features = out_features
+        w_shape = (in_features, out_features)
+        self.w = Tensor(shape=w_shape, requires_grad=True, stores_grad=True)
+        if bias:
+            b_shape = (1, out_features)
+            self.b = Tensor(shape=b_shape, requires_grad=True, stores_grad=True)
+        self.init_value = False
+
+    def get_params(self):
+        if hasattr(self, 'b'):
+            return (self.w, self.b)
+        else:
+            return self.w
+
+    def __call__(self, x, flag=True):
+        assert type(flag) is bool, 'flag can only be bool.'
+        self.flag = flag
+        if self.init_value is False:
+            std = math.sqrt(2.0 / (self.in_features + self.out_features))
+            self.w.gaussian(0.0, std)
```

Comment:
put the init steps into the __init__()?

Reply:
i think its may hard to do so unless realize an alternative setup method for CLayer by sending kernel size.
if put the init into init(), i am not sure if the parameters can connect with CLayer, it must refer the setup codes for Clayer to find how to setup CLayer from outside.

Reply:
can easily put init steps into init() here. i wrongly think that you are indicating conv2d as i just look part of codes. and the last comment is for conv2d.